SqlSetup: Working SqlServerDsc SQL 2019 configuration initially fails on SQL 2022 then passes

### Problem description

Greetings community, I use Azure Automation State Configuration to configure SQL Server on Azure VMs. I am testing a Windows Server 2019 / SQL Server 2019 configuration on Windows Server 2022 / SQL Server 2022. The configuration is eventually applied successfully, but it takes much longer because of new failures on SQL 2022. 

The errors in the log indicate connection failures to the SQL Server instance, so for some reason SqlServerDsc is unable to connect to the default SQL Server instance when DSC configuration is first applied. I never saw these initial failures on SQL 2019. I would like to get rid of those initial failures to bring down the amount of time needed for DSC to complete configuration.

Any help would be much appreciated.

### Verbose logs

```text
{
    "Exception":  {
                      "Message":  "PowerShell DSC resource DSC_SqlLogin  failed to execute Test-TargetResource functionality with error message: System.InvalidOperationException: Failed to connect to SQL instance \u0027mssqlwin1\u0027. (SQLCOMMON0019) ---\u003e System.Management.Automation.MethodInvocationException: Exception calling \"Connect\" with \"0\" argument(s): \"Failed to connect to server mssqlwin1.\" ---\u003e Microsoft.SqlServer.Management.Common.ConnectionFailureException: Failed to connect to server mssqlwin1. ---\u003e System.Data.SqlClient.SqlException: A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: Named Pipes Provider, error: 40 - Could not open a connection to SQL Server) ---\u003e System.ComponentModel.Win32Exception: The system cannot find the file specified\r\n   --- End of inner exception stack trace ---\r\n   at System.Data.SqlClient.SqlInternalConnectionTds..ctor(DbConnectionPoolIdentity identity, SqlConnectionString connectionOptions, SqlCredential credential, Object providerInfo, String newPassword, SecureString newSecurePassword, Boolean redirectedUserInstance, SqlConnectionString userConnectionOptions, SessionData reconnectSessionData, DbConnectionPool pool, String accessToken, Boolean applyTransientFaultHandling, SqlAuthenticationProviderManager sqlAuthProviderManager)\r\n   at System.Data.SqlClient.SqlConnectionFactory.CreateConnection(DbConnectionOptions options, DbConnectionPoolKey poolKey, Object poolGroupProviderInfo, DbConnectionPool pool, DbConnection owningConnection, DbConnectionOptions userOptions)\r\n   at System.Data.ProviderBase.DbConnectionFactory.CreatePooledConnection(DbConnectionPool pool, DbConnection owningObject, DbConnectionOptions options, DbConnectionPoolKey poolKey, DbConnectionOptions userOptions)\r\n   at System.Data.ProviderBase.DbConnectionPool.CreateObject(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)\r\n   at System.Data.ProviderBase.DbConnectionPool.UserCreateRequest(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)\r\n   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal\u0026 connection)\r\n   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal\u0026 connection)\r\n   at System.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal\u0026 connection)\r\n   at System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)\r\n   at System.Data.SqlClient.SqlConnection.TryOpenInner(TaskCompletionSource`1 retry)\r\n   at System.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry)\r\n   at System.Data.SqlClient.SqlConnection.Open()\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.InternalConnectImpl()\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.InternalConnect()\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.Connect()\r\n   --- End of inner exception stack trace ---\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.Connect()\r\n   at CallSite.Target(Closure , CallSite , Object )\r\n   --- End of inner exception stack trace ---\r\n   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)\r\n   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)\r\n   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)\r\n   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)\r\n   --- End of inner exception stack trace --- ",
                      "Data":  {

                               },
                      "InnerException":  {
                                             "SerializedRemoteException":  "System.Exception: System.InvalidOperationException: Failed to connect to SQL instance \u0027mssqlwin1\u0027. (SQLCOMMON0019) ---\u003e System.Management.Automation.MethodInvocationException: Exception calling \"Connect\" with \"0\" argument(s): \"Failed to connect to server mssqlwin1.\" ---\u003e Microsoft.SqlServer.Management.Common.ConnectionFailureException: Failed to connect to server mssqlwin1. ---\u003e System.Data.SqlClient.SqlException: A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: Named Pipes Provider, error: 40 - Could not open a connection to SQL Server) ---\u003e System.ComponentModel.Win32Exception: The system cannot find the file specified\r\n   --- End of inner exception stack trace ---\r\n   at System.Data.SqlClient.SqlInternalConnectionTds..ctor(DbConnectionPoolIdentity identity, SqlConnectionString connectionOptions, SqlCredential credential, Object providerInfo, String newPassword, SecureString newSecurePassword, Boolean redirectedUserInstance, SqlConnectionString userConnectionOptions, SessionData reconnectSessionData, DbConnectionPool pool, String accessToken, Boolean applyTransientFaultHandling, SqlAuthenticationProviderManager sqlAuthProviderManager)\r\n   at System.Data.SqlClient.SqlConnectionFactory.CreateConnection(DbConnectionOptions options, DbConnectionPoolKey poolKey, Object poolGroupProviderInfo, DbConnectionPool pool, DbConnection owningConnection, DbConnectionOptions userOptions)\r\n   at System.Data.ProviderBase.DbConnectionFactory.CreatePooledConnection(DbConnectionPool pool, DbConnection owningObject, DbConnectionOptions options, DbConnectionPoolKey poolKey, DbConnectionOptions userOptions)\r\n   at System.Data.ProviderBase.DbConnectionPool.CreateObject(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)\r\n   at System.Data.ProviderBase.DbConnectionPool.UserCreateRequest(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)\r\n   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal\u0026 connection)\r\n   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal\u0026 connection)\r\n   at System.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal\u0026 connection)\r\n   at System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)\r\n   at System.Data.SqlClient.SqlConnection.TryOpenInner(TaskCompletionSource`1 retry)\r\n   at System.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry)\r\n   at System.Data.SqlClient.SqlConnection.Open()\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.InternalConnectImpl()\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.InternalConnect()\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.Connect()\r\n   --- End of inner exception stack trace ---\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.Connect()\r\n   at CallSite.Target(Closure , CallSite , Object )\r\n   --- End of inner exception stack trace ---\r\n   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)\r\n   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)\r\n   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)\r\n   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)\r\n   --- End of inner exception stack trace ---",
                                             "SerializedRemoteInvocationInfo":  "System.Management.Automation.InvocationInfo",
                                             "ErrorRecord":  "System.InvalidOperationException: Failed to connect to SQL instance \u0027mssqlwin1\u0027. (SQLCOMMON0019) ---\u003e System.Management.Automation.MethodInvocationException: Exception calling \"Connect\" with \"0\" argument(s): \"Failed to connect to server mssqlwin1.\" ---\u003e Microsoft.SqlServer.Management.Common.ConnectionFailureException: Failed to connect to server mssqlwin1. ---\u003e System.Data.SqlClient.SqlException: A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: Named Pipes Provider, error: 40 - Could not open a connection to SQL Server) ---\u003e System.ComponentModel.Win32Exception: The system cannot find the file specified\r\n   --- End of inner exception stack trace ---\r\n   at System.Data.SqlClient.SqlInternalConnectionTds..ctor(DbConnectionPoolIdentity identity, SqlConnectionString connectionOptions, SqlCredential credential, Object providerInfo, String newPassword, SecureString newSecurePassword, Boolean redirectedUserInstance, SqlConnectionString userConnectionOptions, SessionData reconnectSessionData, DbConnectionPool pool, String accessToken, Boolean applyTransientFaultHandling, SqlAuthenticationProviderManager sqlAuthProviderManager)\r\n   at System.Data.SqlClient.SqlConnectionFactory.CreateConnection(DbConnectionOptions options, DbConnectionPoolKey poolKey, Object poolGroupProviderInfo, DbConnectionPool pool, DbConnection owningConnection, DbConnectionOptions userOptions)\r\n   at System.Data.ProviderBase.DbConnectionFactory.CreatePooledConnection(DbConnectionPool pool, DbConnection owningObject, DbConnectionOptions options, DbConnectionPoolKey poolKey, DbConnectionOptions userOptions)\r\n   at System.Data.ProviderBase.DbConnectionPool.CreateObject(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)\r\n   at System.Data.ProviderBase.DbConnectionPool.UserCreateRequest(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)\r\n   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal\u0026 connection)\r\n   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal\u0026 connection)\r\n   at System.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal\u0026 connection)\r\n   at System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)\r\n   at System.Data.SqlClient.SqlConnection.TryOpenInner(TaskCompletionSource`1 retry)\r\n   at System.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry)\r\n   at System.Data.SqlClient.SqlConnection.Open()\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.InternalConnectImpl()\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.InternalConnect()\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.Connect()\r\n   --- End of inner exception stack trace ---\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.Connect()\r\n   at CallSite.Target(Closure , CallSite , Object )\r\n   --- End of inner exception stack trace ---\r\n   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)\r\n   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)\r\n   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)\r\n   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)\r\n   --- End of inner exception stack trace ---",
                                             "WasThrownFromThrowStatement":  false,
                                             "Message":  "System.InvalidOperationException: Failed to connect to SQL instance \u0027mssqlwin1\u0027. (SQLCOMMON0019) ---\u003e System.Management.Automation.MethodInvocationException: Exception calling \"Connect\" with \"0\" argument(s): \"Failed to connect to server mssqlwin1.\" ---\u003e Microsoft.SqlServer.Management.Common.ConnectionFailureException: Failed to connect to server mssqlwin1. ---\u003e System.Data.SqlClient.SqlException: A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: Named Pipes Provider, error: 40 - Could not open a connection to SQL Server) ---\u003e System.ComponentModel.Win32Exception: The system cannot find the file specified\r\n   --- End of inner exception stack trace ---\r\n   at System.Data.SqlClient.SqlInternalConnectionTds..ctor(DbConnectionPoolIdentity identity, SqlConnectionString connectionOptions, SqlCredential credential, Object providerInfo, String newPassword, SecureString newSecurePassword, Boolean redirectedUserInstance, SqlConnectionString userConnectionOptions, SessionData reconnectSessionData, DbConnectionPool pool, String accessToken, Boolean applyTransientFaultHandling, SqlAuthenticationProviderManager sqlAuthProviderManager)\r\n   at System.Data.SqlClient.SqlConnectionFactory.CreateConnection(DbConnectionOptions options, DbConnectionPoolKey poolKey, Object poolGroupProviderInfo, DbConnectionPool pool, DbConnection owningConnection, DbConnectionOptions userOptions)\r\n   at System.Data.ProviderBase.DbConnectionFactory.CreatePooledConnection(DbConnectionPool pool, DbConnection owningObject, DbConnectionOptions options, DbConnectionPoolKey poolKey, DbConnectionOptions userOptions)\r\n   at System.Data.ProviderBase.DbConnectionPool.CreateObject(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)\r\n   at System.Data.ProviderBase.DbConnectionPool.UserCreateRequest(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)\r\n   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal\u0026 connection)\r\n   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal\u0026 connection)\r\n   at System.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal\u0026 connection)\r\n   at System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)\r\n   at System.Data.SqlClient.SqlConnection.TryOpenInner(TaskCompletionSource`1 retry)\r\n   at System.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry)\r\n   at System.Data.SqlClient.SqlConnection.Open()\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.InternalConnectImpl()\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.InternalConnect()\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.Connect()\r\n   --- End of inner exception stack trace ---\r\n   at Microsoft.SqlServer.Management.Common.ConnectionManager.Connect()\r\n   at CallSite.Target(Closure , CallSite , Object )\r\n   --- End of inner exception stack trace ---\r\n   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)\r\n   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)\r\n   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)\r\n   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)\r\n   --- End of inner exception stack trace ---",
                                             "Data":  "System.Collections.ListDictionaryInternal",
                                             "InnerException":  null,
                                             "TargetSite":  "Void EndInvoke()",
                                             "StackTrace":  "   at System.Management.Automation.Runspaces.AsyncResult.EndInvoke()\r\n   at System.Management.Automation.PowerShell.CoreInvokeRemoteHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)\r\n   at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)\r\n   at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)\r\n   at Microsoft.PowerShell.DesiredStateConfiguration.Internal.ResourceProviderAdapter.ExecuteCommand(PowerShell powerShell, ResourceModuleInfo resInfo, String operationCmd, List`1 acceptedProperties, CimInstance nonResourcePropeties, CimInstance resourceConfiguration, LCMDebugMode debugMode, PSInvocationSettings pSInvocationSettings, UInt32\u0026 resultStatusHandle, Collection`1\u0026 result, ErrorRecord\u0026 errorRecord, PSModuleInfo localRunSpaceModuleInfo)",
                                             "HelpLink":  null,
                                             "Source":  "System.Management.Automation",
                                             "HResult":  -2146233087
                                         },
                      "TargetSite":  null,
                      "StackTrace":  null,
                      "HelpLink":  null,
                      "Source":  null,
                      "HResult":  -2146233079
                  },
    "TargetObject":  null,
    "CategoryInfo":  {
                         "Category":  7,
                         "Activity":  "",
                         "Reason":  "InvalidOperationException",
                         "TargetName":  "",
                         "TargetType":  ""
                     },
    "FullyQualifiedErrorId":  "ProviderOperationExecutionFailure",
    "ErrorDetails":  null,
    "InvocationInfo":  null,
    "ScriptStackTrace":  null,
    "PipelineIterationInfo":  [

                              ]
}
```


### DSC configuration

```powershell
configuration MssqlVmConfig {
    param (
        [Parameter(Mandatory = $true)]
        [String]$ComputerName
    )

    Import-DscResource -ModuleName 'PSDscResources'
    Import-DscResource -ModuleName 'xDSCDomainjoin'
    Import-DscResource -ModuleName 'NetworkingDsc'
    Import-DscResource -ModuleName 'SqlServerDsc'
    Import-DscResource -ModuleName 'ActiveDirectoryDsc'
    
    $domain = Get-AutomationVariable -Name 'adds_domain_name'
    $localAdminCredential = Get-AutomationPSCredential 'bootstrapadmin'
    $domainAdminCredential = Get-AutomationPSCredential 'domainadmin'
    $domainAdminShortCredential = Get-AutomationPSCredential 'domainadminshort'

    node $ComputerName {
        SqlScriptQuery 'EnableSa' {
            InstanceName = 'MSSQLSERVER'
            GetQuery = @'
SELECT [is_disabled] FROM [master].[sys].[sql_logins] WHERE ([name] = 'sa') FOR JSON AUTO;
GO
'@
            TestQuery = @'
IF (SELECT [is_disabled] FROM [master].[sys].[sql_logins] WHERE ([name] = 'sa')) = 1 
BEGIN
    RAISERROR ('sa login is disabled.', 16, 1 ) ;
END
ELSE
BEGIN
    PRINT 'sa login is enabled' ;
END ;
GO
'@
            SetQuery = @'
ALTER LOGIN sa ENABLE ;
GO
ALTER LOGIN sa WITH PASSWORD = '$(UsernameSecret)' ;
GO
USE [master]
GO
EXEC xp_instance_regwrite N'HKEY_LOCAL_MACHINE', 
     N'Software\Microsoft\MSSQLServer\MSSQLServer',
     N'LoginMode', REG_DWORD, 2 ;
GO
'@
            Variable = @(
                ('UsernameSecret={0}' -f $localAdminCredential.GetNetworkCredential().Password)
            )

            QueryTimeout = 30
            PSDscRunAsCredential = $localAdminCredential
        }

        xDSCDomainjoin 'JoinDomain' {
            Domain = $domain
            Credential = $domainAdminCredential
            DependsOn = '[SqlScriptQuery]EnableSa'
        }

        Firewall 'MssqlFirewallRule' {
            Name = 'MssqlFirewallRule'
            DisplayName = 'Microsoft SQL Server database engine.'
            Ensure = 'Present'
            Enabled = 'True'
            Profile = ('Domain', 'Private')
            Direction = 'InBound'
            LocalPort = ('1433')
            Protocol = 'TCP'
            DependsOn = '[xDSCDomainjoin]JoinDomain'
        }

        SqlLogin 'DomainAdmin' {
            Name = $domainAdminShortCredential.UserName
            LoginType = 'WindowsUser'
            InstanceName = 'MSSQLSERVER'
            Ensure = 'Present'
            DependsOn = '[xDSCDomainjoin]JoinDomain'
            PSDscRunAsCredential = $localAdminCredential
        }

        SqlRole 'sysadmin' {
            ServerRoleName = 'sysadmin'
            MembersToInclude = $domainAdminShortCredential.UserName
            InstanceName = 'MSSQLSERVER'
            Ensure = 'Present'
            DependsOn = '[SqlLogin]DomainAdmin'
            PSDscRunAsCredential = $localAdminCredential
        }

        WindowsFeature 'RSAT-AD-PowerShell' {
            Name = 'RSAT-AD-PowerShell'
            Ensure = 'Present'
            DependsOn = '[xDSCDomainjoin]JoinDomain'            
        }

        ADGroup 'DatabaseServers' {
            GroupName = 'DatabaseServers'
            GroupScope = 'Global'
            Category = 'Security'
            MembersToInclude = "$ComputerName$"
            Credential = $domainAdminCredential
            Ensure = 'Present'
            DependsOn = '[WindowsFeature]RSAT-AD-PowerShell'            
        }
    }
}
```


### Suggested solution

Add some new checks to SqlServerDsc to make sure the SQL Server 2022 service is running before attempting to connect and apply the configuration.

### SQL Server edition and version

```text
SQL Server 2022
```


### SQL Server PowerShell modules

```text
SQLPS 16.0
```


### Operating system

```text
Windows Server 2022
```


### PowerShell version

```text
5.1
```


### SqlServerDsc version

```text
16.0.0
```

@johlju I continued looking into this.  The SMO ServerConnection Class has a property for [ConnectTimeout](https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.management.common.connectionsettings.connecttimeout?view=sql-smo-160#microsoft-sqlserver-management-common-connectionsettings-connecttimeout).  This is separate than that of the statement timeout and utilizing this would be rather trivial.  The Connect() method will just block until it is successful or exceeds the timeout.  I've implemented this and moved a few things around and it works to get past the Connect() method however I'm seeing the following message now

```
VERBOSE: [2023-03-13 21:37:52Z] [VERBOSE] [mysqlserver01]:                            
[[SqlSetup]InstallDefaultInstance::[InstallSqlServer]Install_SQL_Server] Connecting as current user 'NT 
AUTHORITY\SYSTEM' using integrated security. (SQLCOMMON0054)
VERBOSE: [2023-03-13 21:38:54Z] [VERBOSE] [mysqlserver01]: LCM:  [ End    Test     ]  
[[SqlSetup]InstallDefaultInstance::[InstallSqlServer]Install_SQL_Server]  in 70.0140 seconds.
VERBOSE: [2023-03-13 21:38:54Z] Settings handler status to 'transitioning' 
(C:\Packages\Plugins\Microsoft.Powershell.DSC\2.83.5\Status\0.status)
VERBOSE: [2023-03-13 21:38:54Z] [ERROR] PowerShell DSC resource DSC_SqlSetup  failed to execute Test-TargetResource 
functionality with error message: The running command stopped because the preference variable "ErrorActionPreference" 
or common parameter is set to Stop: System.InvalidOperationException: Failed to connect to SQL instance 
'mysqlserver01'. (SQLCOMMON0019) ---> System.Exception: System.InvalidOperationException: The SQL instance 
'mysqlserver01' was expected to have the status 'Online', but had status 'Unknown'. (SQLCOMMON0071)

   --- End of inner exception stack trace --- 
```

I'm not sure why the connect would work but the status from the SMO Server object would still be null.  In my testing the Status is updated to 'Online' almost instantly as the SQL Service starts, which happens automatically without needing to use the Refresh() method.  Though I will say the opposite is not true.  The status property will still show 'Online' even if you stop the service, which is interesting though not necessarily related to conversation at hand, I just wanted to point it out.

I'm going to keep looking into it but wanted to put some updates in here as I continue.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -38,6 +38,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - SqlSetup
   - Update to support checking non-supported features using the command
     `SqlDscIsSupportedFeature` ([issue #1872](https://github.com/dsccommunity/SqlServerDsc/issues/1872)).
+- Update SqlServerDsc.Common Module:
+  - `Connect-SQL` - Function will now wait for the SMO Status property to be
+    'Online' or throw an exception if time exceeds the statement timeout.
 
 ## [16.2.0] - 2023-04-10
 

--- a/source/Modules/SqlServerDsc.Common/SqlServerDsc.Common.psm1
+++ b/source/Modules/SqlServerDsc.Common/SqlServerDsc.Common.psm1
@@ -562,6 +562,7 @@ function Connect-SQL
             report 'Online' immediately eventhough the Connect() was successful.  The loop is to
             ensure the SMO's Status property was been updated.
         #>
+        $sleepInSeconds = 2
         do
         {
             $instanceStatus = $sqlServerObject.Status
@@ -580,11 +581,11 @@ function Connect-SQL
                 break
             }
 
-            Write-Verbose -Message (
-                $script:localizedData.WaitForDatabaseEngineInstanceStatus -f $instanceStatus, $onlineStatus
-            ) -Verbose
+            Write-Debug -Message (
+                $script:localizedData.WaitForDatabaseEngineInstanceStatus -f $instanceStatus, $onlineStatus, $sleepInSeconds
+            )
 
-            Start-Sleep -Seconds 2
+            Start-Sleep -Seconds $sleepInSeconds
             $sqlServerObject.Refresh()
         } while ($connectTimer.Elapsed.TotalSeconds -lt $StatementTimeout)
 

--- a/source/Modules/SqlServerDsc.Common/SqlServerDsc.Common.psm1
+++ b/source/Modules/SqlServerDsc.Common/SqlServerDsc.Common.psm1
@@ -511,6 +511,7 @@ function Connect-SQL
     $sqlConnectionContext = $sqlServerObject.ConnectionContext
     $sqlConnectionContext.ServerInstance = $databaseEngineInstance
     $sqlConnectionContext.StatementTimeout = $StatementTimeout
+    $sqlConnectionContext.ConnectTimeout = $StatementTimeout
     $sqlConnectionContext.ApplicationName = 'SqlServerDsc'
 
     if ($PSCmdlet.ParameterSetName -eq 'SqlServer')
@@ -551,15 +552,43 @@ function Connect-SQL
 
     try
     {
+        $onlineStatus = 'Online'
+        $connectTimer = [System.Diagnostics.StopWatch]::StartNew()
         $sqlConnectionContext.Connect()
 
-        $instanceStatus = $sqlServerObject.Status
-
-        if ($instanceStatus)
+        <#
+            The addition of the ConnetTimeout property to the ConnectionContext will force the
+            Connect() method to block until successful.  THe SMO object's Status property may not
+            report 'Online' immediately eventhough the Connect() was successful.  The loop is to
+            ensure the SMO's Status property was been updated.
+        #>
+        do
         {
-            # Property Status is of type Enum ServerStatus, we return the string equivalent.
-            $instanceStatus = $instanceStatus.ToString()
-        }
+            $instanceStatus = $sqlServerObject.Status
+            if ([System.String]::IsNullOrEmpty($instanceStatus))
+            {
+                $instanceStatus = 'Unknown'
+            }
+            else
+            {
+                # Property Status is of type Enum ServerStatus, we return the string equivalent.
+                $instanceStatus = $instanceStatus.ToString()
+            }
+
+            if ($instanceStatus -eq $onlineStatus)
+            {
+                break
+            }
+
+            Write-Verbose -Message (
+                $script:localizedData.WaitForDatabaseEngineInstanceStatus -f $instanceStatus, $onlineStatus
+            ) -Verbose
+
+            Start-Sleep -Seconds 2
+            $sqlServerObject.Refresh()
+        } while ($connectTimer.Elapsed.TotalSeconds -lt $StatementTimeout)
+
+        $connectTimer.Stop()
 
         if ($instanceStatus -match '^Online$')
         {
@@ -571,11 +600,6 @@ function Connect-SQL
         }
         else
         {
-            if ([System.String]::IsNullOrEmpty($instanceStatus))
-            {
-                $instanceStatus = 'Unknown'
-            }
-
             $errorMessage = $script:localizedData.DatabaseEngineInstanceNotOnline -f @(
                 $databaseEngineInstance,
                 $instanceStatus

--- a/source/Modules/SqlServerDsc.Common/SqlServerDsc.Common.psm1
+++ b/source/Modules/SqlServerDsc.Common/SqlServerDsc.Common.psm1
@@ -589,8 +589,6 @@ function Connect-SQL
             $sqlServerObject.Refresh()
         } while ($connectTimer.Elapsed.TotalSeconds -lt $StatementTimeout)
 
-        $connectTimer.Stop()
-
         if ($instanceStatus -match '^Online$')
         {
             Write-Verbose -Message (
@@ -645,6 +643,7 @@ function Connect-SQL
     }
     finally
     {
+        $connectTimer.Stop()
         <#
             Connect will ensure we actually can connect, but we need to disconnect
             from the session so we don't have anything hanging. If we need run a

--- a/source/Modules/SqlServerDsc.Common/en-US/SqlServerDsc.Common.strings.psd1
+++ b/source/Modules/SqlServerDsc.Common/en-US/SqlServerDsc.Common.strings.psd1
@@ -44,4 +44,5 @@ ConvertFrom-StringData @'
     FailedToLoadAssembly = Failed to load the assembly '{0}'. (SQLCOMMON0069)
     FailedToObtainServerInstance = Failed to obtain a SQL Server instance with name '{0}' on server '{1}'. Ensure the SQL Server instance exists on the server and that the 'SQLServer' module references a version of the 'Microsoft.SqlServer.Management.Smo.Wmi' library that supports the version of the SQL Server instance. (SQLCOMMON0070)
     DatabaseEngineInstanceNotOnline = The SQL instance '{0}' was expected to have the status 'Online', but had status '{1}'. (SQLCOMMON0071)
+    WaitForDatabaseEngineInstanceStatus = The SQL instance status is '{0}' waiting for '{1}'.  (SQLCOMMON0072)"
 '@

--- a/source/Modules/SqlServerDsc.Common/en-US/SqlServerDsc.Common.strings.psd1
+++ b/source/Modules/SqlServerDsc.Common/en-US/SqlServerDsc.Common.strings.psd1
@@ -44,5 +44,5 @@ ConvertFrom-StringData @'
     FailedToLoadAssembly = Failed to load the assembly '{0}'. (SQLCOMMON0069)
     FailedToObtainServerInstance = Failed to obtain a SQL Server instance with name '{0}' on server '{1}'. Ensure the SQL Server instance exists on the server and that the 'SQLServer' module references a version of the 'Microsoft.SqlServer.Management.Smo.Wmi' library that supports the version of the SQL Server instance. (SQLCOMMON0070)
     DatabaseEngineInstanceNotOnline = The SQL instance '{0}' was expected to have the status 'Online', but had status '{1}'. (SQLCOMMON0071)
-    WaitForDatabaseEngineInstanceStatus = The SQL instance status is '{0}' waiting for '{1}'.  (SQLCOMMON0072)"
+    WaitForDatabaseEngineInstanceStatus = The SQL instance status is '{0}' expected '{1}', waiting {2} seconds.  (SQLCOMMON0072)
 '@

--- a/source/Modules/SqlServerDsc.Common/sv-SE/SqlServerDsc.Common.strings.psd1
+++ b/source/Modules/SqlServerDsc.Common/sv-SE/SqlServerDsc.Common.strings.psd1
@@ -50,4 +50,5 @@ ConvertFrom-StringData @'
     FailedToLoadAssembly = Failed to load the assembly '{0}'. (SQLCOMMON0069)
     FailedToObtainServerInstance = Failed to obtain a SQL Server instance with name '{0}' on server '{1}'. Ensure the SQL Server instance exists on the server and that the 'SQLServer' module references a version of the 'Microsoft.SqlServer.Management.Smo.Wmi' library that supports the version of the SQL Server instance. (SQLCOMMON0070)
     DatabaseEngineInstanceNotOnline = The SQL instance '{0}' was expected to have the status 'Online', but had status '{1}'. (SQLCOMMON0071)
+    WaitForDatabaseEngineInstanceStatus = The SQL instance status is '{0}' waiting for '{1}'.  (SQLCOMMON0072)"
 '@

--- a/source/Modules/SqlServerDsc.Common/sv-SE/SqlServerDsc.Common.strings.psd1
+++ b/source/Modules/SqlServerDsc.Common/sv-SE/SqlServerDsc.Common.strings.psd1
@@ -50,5 +50,5 @@ ConvertFrom-StringData @'
     FailedToLoadAssembly = Failed to load the assembly '{0}'. (SQLCOMMON0069)
     FailedToObtainServerInstance = Failed to obtain a SQL Server instance with name '{0}' on server '{1}'. Ensure the SQL Server instance exists on the server and that the 'SQLServer' module references a version of the 'Microsoft.SqlServer.Management.Smo.Wmi' library that supports the version of the SQL Server instance. (SQLCOMMON0070)
     DatabaseEngineInstanceNotOnline = The SQL instance '{0}' was expected to have the status 'Online', but had status '{1}'. (SQLCOMMON0071)
-    WaitForDatabaseEngineInstanceStatus = The SQL instance status is '{0}' waiting for '{1}'.  (SQLCOMMON0072)"
+    WaitForDatabaseEngineInstanceStatus = The SQL instance status is '{0}' expected '{1}', waiting {2} seconds.  (SQLCOMMON0072)
 '@

--- a/tests/Unit/SqlServerDsc.Common.Tests.ps1
+++ b/tests/Unit/SqlServerDsc.Common.Tests.ps1
@@ -2364,6 +2364,7 @@ Describe 'SqlServerDsc.Common\Connect-SQL' -Tag 'ConnectSql' {
                         Add-Member -MemberType NoteProperty -Name ConnectAsUserPassword -Value '' -PassThru |
                         Add-Member -MemberType NoteProperty -Name ConnectAsUserName -Value '' -PassThru |
                         Add-Member -MemberType NoteProperty -Name StatementTimeout -Value 600 -PassThru |
+                        Add-Member -MemberType NoteProperty -Name ConnectTimeout -Value 600 -PassThru |
                         Add-Member -MemberType NoteProperty -Name ApplicationName -Value 'SqlServerDsc' -PassThru |
                         Add-Member -MemberType ScriptMethod -Name Disconnect -Value {
                             return $true
@@ -2622,6 +2623,7 @@ Describe 'SqlServerDsc.Common\Connect-SQL' -Tag 'ConnectSql' {
                                 Add-Member -MemberType NoteProperty -Name ConnectAsUserPassword -Value '' -PassThru |
                                 Add-Member -MemberType NoteProperty -Name ConnectAsUserName -Value '' -PassThru |
                                 Add-Member -MemberType NoteProperty -Name StatementTimeout -Value 600 -PassThru |
+                                Add-Member -MemberType NoteProperty -Name ConnectTimeout -Value 600 -PassThru |
                                 Add-Member -MemberType NoteProperty -Name ApplicationName -Value 'SqlServerDsc' -PassThru |
                                 Add-Member -MemberType ScriptMethod -Name Disconnect -Value {
                                     return $true
@@ -2668,6 +2670,7 @@ Describe 'SqlServerDsc.Common\Connect-SQL' -Tag 'ConnectSql' {
                                 Add-Member -MemberType NoteProperty -Name ConnectAsUserPassword -Value '' -PassThru |
                                 Add-Member -MemberType NoteProperty -Name ConnectAsUserName -Value '' -PassThru |
                                 Add-Member -MemberType NoteProperty -Name StatementTimeout -Value 600 -PassThru |
+                                Add-Member -MemberType NoteProperty -Name ConnectTimeout -Value 600 -PassThru |
                                 Add-Member -MemberType NoteProperty -Name ApplicationName -Value 'SqlServerDsc' -PassThru |
                                 Add-Member -MemberType ScriptMethod -Name Disconnect -Value {
                                     return $true

